File: app/src/main/java/com/littlefencer/app/logic/detectors/FlungeDetector.kt
package com.littlefencer.app.logic.detectors

import com.littlefencer.app.logic.ActionDetector
import com.littlefencer.app.logic.ActionQuality
import com.littlefencer.app.logic.ActionResult
import com.littlefencer.app.logic.PoseFrame
import com.littlefencer.app.logic.SaberAction
import com.littlefencer.app.utils.GeometryUtils
import com.littlefencer.app.utils.MediaPipeLandmarks

/**
 * FlungeDetector - Detects the Flunge (飞弓步) - saber's replacement for the banned fleche.
 * 
 * Biomechanics:
 * - A "flying lunge" - essentially an exaggerated lunge with explosive forward momentum
 * - Back foot may leave ground but MUST NOT pass the front foot (rule)
 * - Body fully commits forward with maximum arm extension
 * - Very high forward velocity compared to normal lunge
 * 
 * Detection Strategy:
 * 1. Arm extends first (attacking principle)
 * 2. Detect explosive forward velocity (much higher than normal lunge)
 * 3. Check back leg leaves ground briefly (optional but common)
 * 4. Verify back foot never passes front foot (rule compliance)
 */
class FlungeDetector : ActionDetector {

    override val targetAction = SaberAction.FLUNGE

    private var isFlunging = false
    private var flungeStartTime = 0L
    private var peakVelocity = 0f
    
    companion object {
        // Flunge requires higher velocity than normal lunge
        private const val FLUNGE_VELOCITY_THRESHOLD = 0.8f  // vs 0.5f for lunge
        
        // Arm must be extended first
        private const val ARM_EXTENDED_THRESHOLD = 0.25f
        
        // Maximum duration
        private const val MAX_FLUNGE_DURATION = 600L  // Faster than normal lunge
        
        // Back leg should be nearly straight
        private const val BACK_KNEE_MIN_STRAIGHT = 155f
    }

    override fun detect(current: PoseFrame, history: List<PoseFrame>): ActionResult {
        if (history.size < 5) return ActionResult.None

        val landmarks = current.landmarks
        if (landmarks.size < 33) return ActionResult.None

        // Extract landmarks
        val leftShoulder = landmarks[MediaPipeLandmarks.LEFT_SHOULDER]
        val rightShoulder = landmarks[MediaPipeLandmarks.RIGHT_SHOULDER]
        val leftHip = landmarks[MediaPipeLandmarks.LEFT_HIP]
        val rightHip = landmarks[MediaPipeLandmarks.RIGHT_HIP]
        val leftWrist = landmarks[MediaPipeLandmarks.LEFT_WRIST]
        val rightWrist = landmarks[MediaPipeLandmarks.RIGHT_WRIST]
        val leftKnee = landmarks[MediaPipeLandmarks.LEFT_KNEE]
        val rightKnee = landmarks[MediaPipeLandmarks.RIGHT_KNEE]
        val leftAnkle = landmarks[MediaPipeLandmarks.LEFT_ANKLE]
        val rightAnkle = landmarks[MediaPipeLandmarks.RIGHT_ANKLE]

        val shoulderMidpoint = GeometryUtils.midpoint(leftShoulder, rightShoulder)
        val ankleMidpoint = GeometryUtils.midpoint(leftAnkle, rightAnkle)
        val bodyHeight = GeometryUtils.distance(shoulderMidpoint, ankleMidpoint)
        
        // Determine front side
        val isLeftFront = leftWrist.x > rightWrist.x
        val frontWrist = if (isLeftFront) leftWrist else rightWrist
        val frontShoulder = if (isLeftFront) leftShoulder else rightShoulder
        val frontAnkle = if (isLeftFront) leftAnkle else rightAnkle
        val backAnkle = if (isLeftFront) rightAnkle else leftAnkle
        
        // Arm extension
        val armExtension = if (bodyHeight > 0) {
            GeometryUtils.distance(frontShoulder, frontWrist) / bodyHeight
        } else 0f
        
        // Back knee angle
        val backKneeAngle = if (isLeftFront) {
            GeometryUtils.angleBetweenPoints(rightHip, rightKnee, rightAnkle)
        } else {
            GeometryUtils.angleBetweenPoints(leftHip, leftKnee, leftAnkle)
        }

        // Calculate forward velocity
        val prevFrame = history.last()
        val prevLandmarks = prevFrame.landmarks
        if (prevLandmarks.size < 33) return ActionResult.None
        
        val prevFrontWrist = if (isLeftFront) {
            prevLandmarks[MediaPipeLandmarks.LEFT_WRIST]
        } else {
            prevLandmarks[MediaPipeLandmarks.RIGHT_WRIST]
        }
        
        val deltaTime = current.timeDeltaMs(prevFrame)
        val wristVelocity = if (deltaTime > 0) {
            (frontWrist.x - prevFrontWrist.x) / deltaTime * 1000f
        } else 0f

        // Rule check: back foot must not pass front foot
        val backFootPassed = backAnkle.x > frontAnkle.x

        return if (!isFlunging) {
            // Check for flunge initiation
            if (armExtension > ARM_EXTENDED_THRESHOLD && 
                wristVelocity > FLUNGE_VELOCITY_THRESHOLD &&
                !backFootPassed) {
                
                isFlunging = true
                flungeStartTime = current.timestampMs
                peakVelocity = wristVelocity
                
                ActionResult.InProgress(SaberAction.FLUNGE, 0.6f)
            } else {
                ActionResult.None
            }
        } else {
            val elapsed = current.timestampMs - flungeStartTime
            
            // Track peak velocity
            if (wristVelocity > peakVelocity) {
                peakVelocity = wristVelocity
            }
            
            // Rule violation check
            if (backFootPassed) {
                reset()
                return ActionResult.Completed(
                    action = SaberAction.FLUNGE,
                    quality = ActionQuality.POOR,
                    feedback = "Back foot passed! (Rule violation)"
                )
            }
            
            // Check timeout
            if (elapsed > MAX_FLUNGE_DURATION) {
                reset()
                return ActionResult.None
            }
            
            // Check completion (velocity decreasing, arm extended, back leg straight)
            if (wristVelocity < peakVelocity * 0.3f &&
                armExtension > ARM_EXTENDED_THRESHOLD &&
                backKneeAngle > BACK_KNEE_MIN_STRAIGHT) {
                
                val quality = evaluateQuality(peakVelocity, armExtension, backKneeAngle)
                reset()
                
                ActionResult.Completed(
                    action = SaberAction.FLUNGE,
                    quality = quality,
                    durationMs = elapsed
                )
            } else {
                ActionResult.InProgress(SaberAction.FLUNGE, 0.7f)
            }
        }
    }
    
    private fun evaluateQuality(velocity: Float, armExt: Float, backKnee: Float): ActionQuality {
        val velocityGood = velocity > FLUNGE_VELOCITY_THRESHOLD * 1.5f
        val armGood = armExt > ARM_EXTENDED_THRESHOLD * 1.2f
        val legGood = backKnee > 165f
        
        val score = listOf(velocityGood, armGood, legGood).count { it }
        
        return when (score) {
            3 -> ActionQuality.PERFECT
            2 -> ActionQuality.GOOD
            1 -> ActionQuality.ACCEPTABLE
            else -> ActionQuality.POOR
        }
    }

    override fun reset() {
        isFlunging = false
        flungeStartTime = 0L
        peakVelocity = 0f
    }
}

File: LittleFencer-iOS/LittleFencer/Logic/Detectors/FlungeDetector.swift
import CoreGraphics
import Foundation

/// Detects the Flunge (飞弓步) - saber's replacement for the banned fleche
///
/// Biomechanics:
/// - A "flying lunge" - essentially an exaggerated lunge with explosive forward momentum
/// - Back foot may leave ground but MUST NOT pass the front foot (rule)
/// - Body fully commits forward with maximum arm extension
/// - Very high forward velocity compared to normal lunge
class FlungeDetector: ActionDetector {
    
    let targetAction = SaberAction.flunge
    
    private var isFlunging = false
    private var flungeStartTime: Double = 0
    private var peakVelocity: CGFloat = 0
    
    // MARK: - Thresholds
    
    private let FLUNGE_VELOCITY_THRESHOLD: CGFloat = 0.8
    private let ARM_EXTENDED_THRESHOLD: CGFloat = 0.25
    private let MAX_FLUNGE_DURATION: Double = 600
    private let BACK_KNEE_MIN_STRAIGHT: CGFloat = 155
    
    func detect(current: PoseFrame, history: [PoseFrame]) -> ActionResult {
        guard history.count >= 5, current.landmarks.count >= 33 else {
            return .none
        }
        
        let landmarks = current.landmarks
        
        // Extract landmarks
        let leftShoulder = landmarks[PoseLandmarks.LEFT_SHOULDER]
        let rightShoulder = landmarks[PoseLandmarks.RIGHT_SHOULDER]
        let leftHip = landmarks[PoseLandmarks.LEFT_HIP]
        let rightHip = landmarks[PoseLandmarks.RIGHT_HIP]
        let leftWrist = landmarks[PoseLandmarks.LEFT_WRIST]
        let rightWrist = landmarks[PoseLandmarks.RIGHT_WRIST]
        let leftKnee = landmarks[PoseLandmarks.LEFT_KNEE]
        let rightKnee = landmarks[PoseLandmarks.RIGHT_KNEE]
        let leftAnkle = landmarks[PoseLandmarks.LEFT_ANKLE]
        let rightAnkle = landmarks[PoseLandmarks.RIGHT_ANKLE]
        
        let shoulderMidpoint = GeometryUtils.midpoint(leftShoulder, rightShoulder)
        let ankleMidpoint = GeometryUtils.midpoint(leftAnkle, rightAnkle)
        let bodyHeight = GeometryUtils.distance(shoulderMidpoint, ankleMidpoint)
        
        // Determine front side
        let isLeftFront = leftWrist.x > rightWrist.x
        let frontWrist = isLeftFront ? leftWrist : rightWrist
        let frontShoulder = isLeftFront ? leftShoulder : rightShoulder
        let frontAnkle = isLeftFront ? leftAnkle : rightAnkle
        let backAnkle = isLeftFront ? rightAnkle : leftAnkle
        
        // Arm extension
        let armExtension = bodyHeight > 0 
            ? GeometryUtils.distance(frontShoulder, frontWrist) / bodyHeight 
            : 0
        
        // Back knee angle
        let backKneeAngle: CGFloat
        if isLeftFront {
            backKneeAngle = GeometryUtils.angleBetweenPoints(rightHip, rightKnee, rightAnkle)
        } else {
            backKneeAngle = GeometryUtils.angleBetweenPoints(leftHip, leftKnee, leftAnkle)
        }
        
        // Previous frame for velocity
        guard let prevFrame = history.last,
              prevFrame.landmarks.count >= 33 else {
            return .none
        }
        
        let prevFrontWrist = isLeftFront 
            ? prevFrame.landmarks[PoseLandmarks.LEFT_WRIST]
            : prevFrame.landmarks[PoseLandmarks.RIGHT_WRIST]
        
        let deltaTime = current.timeDeltaMs(prevFrame)
        let wristVelocity: CGFloat = deltaTime > 0 
            ? (frontWrist.x - prevFrontWrist.x) / CGFloat(deltaTime) * 1000 
            : 0
        
        // Rule check: back foot must not pass front foot
        let backFootPassed = backAnkle.x > frontAnkle.x
        
        if !isFlunging {
            // Check for flunge initiation
            if armExtension > ARM_EXTENDED_THRESHOLD &&
               wristVelocity > FLUNGE_VELOCITY_THRESHOLD &&
               !backFootPassed {
                isFlunging = true
                flungeStartTime = current.timestampMs
                peakVelocity = wristVelocity
                return .inProgress(action: .flunge, confidence: 0.6)
            }
            return .none
        } else {
            let elapsed = current.timestampMs - flungeStartTime
            
            // Track peak velocity
            if wristVelocity > peakVelocity {
                peakVelocity = wristVelocity
            }
            
            // Rule violation check
            if backFootPassed {
                reset()
                return .completed(action: .flunge, quality: .poor,
                                 feedback: "Back foot passed! (Rule violation)",
                                 durationMs: elapsed)
            }
            
            // Check timeout
            if elapsed > MAX_FLUNGE_DURATION {
                reset()
                return .none
            }
            
            // Check completion
            if wristVelocity < peakVelocity * 0.3 &&
               armExtension > ARM_EXTENDED_THRESHOLD &&
               backKneeAngle > BACK_KNEE_MIN_STRAIGHT {
                
                let quality = evaluateQuality(armExtension: armExtension,
                                             backKneeAngle: backKneeAngle,
                                             peakVelocity: peakVelocity)
                let feedback = generateFeedback(armExtension: armExtension,
                                               backKneeAngle: backKneeAngle)
                reset()
                return .completed(action: .flunge, quality: quality,
                                 feedback: feedback, durationMs: elapsed)
            }
            
            return .inProgress(action: .flunge, confidence: 0.7)
        }
    }
    
    private func evaluateQuality(armExtension: CGFloat, backKneeAngle: CGFloat, peakVelocity: CGFloat) -> ActionQuality {
        let armGood = armExtension > ARM_EXTENDED_THRESHOLD * 1.2
        let legGood = backKneeAngle > 165
        let velocityGood = peakVelocity > FLUNGE_VELOCITY_THRESHOLD * 1.5
        
        var score = 0
        if armGood { score += 1 }
        if legGood { score += 1 }
        if velocityGood { score += 1 }
        
        switch score {
        case 3: return .perfect
        case 2: return .good
        case 1: return .acceptable
        default: return .poor
        }
    }
    
    private func generateFeedback(armExtension: CGFloat, backKneeAngle: CGFloat) -> String? {
        if armExtension < ARM_EXTENDED_THRESHOLD { return "Arm first!" }
        if backKneeAngle < BACK_KNEE_MIN_STRAIGHT { return "Push back leg!" }
        return nil
    }
    
    func reset() {
        isFlunging = false
        flungeStartTime = 0
        peakVelocity = 0
    }
}
